onnxruntime/test/ir/flatbuffer_utils_test.cc

```diff
@@ -0,0 +1,409 @@
+  // need 128 bytes to write to external data
+
+  // 16-bit. 81 elements so we're 2 bytes past 4 byte alignment
+  initializers.emplace_back(
+      CreateInitializer<int16_t>("tensor_16", ONNX_NAMESPACE::TensorProto_DataType_INT16, {9, 9}));
+
+  // 32-bit, 64 elements
+  initializers.emplace_back(
+      CreateInitializer<float>("tensor_f32", ONNX_NAMESPACE::TensorProto_DataType_FLOAT, {8, 8}));
+
+  // 8-bit. 129 elements so we're 1 byte past 4 or 8 byte alignment
+  initializers.emplace_back(
+      CreateInitializer<uint8_t>("tensor_8", ONNX_NAMESPACE::TensorProto_DataType_UINT8, {3, 43}));
+
+  // 64-bit, 36 elements
+  initializers.emplace_back(
+      CreateInitializer<int64_t>("tensor_64", ONNX_NAMESPACE::TensorProto_DataType_INT64, {6, 6}));
+
+  // small (should not use external)
+  initializers.emplace_back(
+      CreateInitializer<int32_t>("tensor_32_small", ONNX_NAMESPACE::TensorProto_DataType_INT32, {2, 2}));
+
+  return initializers;
+}
+#endif // ENABLE_TRAINING_APIS
+
+template <typename T>
+std::vector<T> ConvertRawDataToTypedVector(ONNX_NAMESPACE::TensorProto initializer) {
+  std::vector<T> data;
+  data.resize(initializer.raw_data().size() / sizeof(T));
+  memcpy(data.data(), initializer.raw_data().data(), initializer.raw_data().size());
+  return data;
+}
+
+#define ASSERT_EQ_FB_TENSORPROTO_VECTORFIELD(EXPECTED, ACTUAL, FIELD) \
+  ASSERT_EQ(EXPECTED.FIELD.size(), ACTUAL.FIELD.size());              \
+  for (int j = 0; j < EXPECTED.FIELD.size(); ++j) {                   \
+    ASSERT_EQ(EXPECTED.FIELD[j], ACTUAL.FIELD[j]);                    \
+  }
+
+}  // namespace
+
+// tests method that loads to tensorproto protobuf (used when loading a checkpoint into an inference model)
+TEST(GraphUtilsTest, ExternalWriteReadWithLoadInitializers) {
+  // create data
+  auto initializers = CreateInitializers();
+
+  flatbuffers::FlatBufferBuilder builder(1024);
+
+  // write
+  std::ofstream output_stream;
+  ExternalDataWriter writer;
+  CreateWriter("ExternalWriteReadBasicTest.bin", output_stream, writer);
+
+  std::vector<flatbuffers::Offset<fbs::Tensor>> fbs_tensors;
+  for (const auto& initializer : initializers) {
+    flatbuffers::Offset<fbs::Tensor> fbs_tensor;
+    ASSERT_STATUS_OK(SaveInitializerOrtFormat(builder, initializer, Path(), fbs_tensor, writer));
+    fbs_tensors.push_back(fbs_tensor);
+  }
+
+  // TODO: might be 844 depending on whether it's 4 byte or 8 byte alignment
+  ASSERT_EQ(output_stream.tellp(), 840) << "Data written to the external file is incorrect.";
+  output_stream.close();
+  ASSERT_TRUE(output_stream.good()) << "Failed to close data file.";
+
+  auto fbs_tensors_offset = builder.CreateVector(fbs_tensors);
+  fbs::test::TestDataBuilder tdb(builder);
+  tdb.add_initializers(fbs_tensors_offset);
+  builder.Finish(tdb.Finish());
+  auto fb_data = builder.GetBufferSpan();
+
+  auto test_data = fbs::test::GetTestData(fb_data.data());
+  auto fbs_tensors2 = test_data->initializers();
```

Comment:
instead of `name` and `name2` you can put variables into an inner scope.
```c++
{
    // ...
    std::vector<flatbuffers::Offset<fbs::Tensor>> fbs_tensors;
    // ...
}

{
    // ...
    auto fbs_tensors = test_data->initializers();
    // ...
}
```

though in this case it might be better to name the first one `fbs_tensor_offsets` since they are different types.

Reply:
Unfortunately they're similar types with different constness, we need to create the builder after fbs_tensors, and we need it to start around to validate the data in it. 

First is std::vector<flatbuffers::Offset<fbs::Tensor>>
Second is const std::vector<flatbuffers::Offset<fbs::Tensor>, flatbuffers::uoffset_t>